agol-publish/NightStocker.py

```diff
@@ -0,0 +1,518 @@
+        groups: list of group names to share layer with (list of strings)
+        tags: comma-separated string of tags (string)
+        description: AGOL description (string)
+        terms_of_use: AGOL terms of use/license info (string)
+        credits: AGOL Credits/Attribution (string)
+        folder: AGOL org's folder to move item to
+    protect: if True, set AGOL flag to prevent item from being deleted
+
+    returns: the published feature layer's itemid
+    '''
+
+    print("uploading")
+    sd_item = gis.content.add({}, data=service_definition)
+
+    #: Publishing
+    print("publishing")
+    published_item = sd_item.publish()
+
+    #: Updating information
+    print("sharing") #: Everyone and groups.
+    published_item.share(everyone=True, org=True, groups=info['groups'])
+    if protect:
+        print("delete protection")
+        published_item.protect(enable=True)
+    # sd_item.protect(enable=True)
+    # print('authoritative')
+    # published_item.content_status = 'authoritative'
+
+    print("updating info")
+    published_item.update(item_properties={
+        'tags': info['tags'],
+        'description': info['description'],
+        'licenseInfo': info['terms_of_use'],
+        'snippet': info['summary'],
+        'accessInformation': info['credits']
+        })
+
+    print('folder')
+    published_item.move(info['folder'])
+    sd_item.move(info['folder'])
+
+    #: Allow Downloads
+    print("downloads")
+    manager = arcgis.features.FeatureLayerCollection.fromitem(published_item).manager
+    manager.update_definition({ 'capabilities': 'Query,Extract' })
+
+    return published_item.itemid
+
+
+def create_service_definition(layer_info, sde_path, temp_dir, project_path, 
+                              map_name, describe):
+    '''Create a service defintion for a layer to be uploaded to AGOL from an
+    SDE using an existing ArcGIS Pro project.
+    
+    Parameters:
+    layer_info: Dictionary of info about the layer to be prepped for upload
+        fc_name: Fully qualified name of the SDE feature class to be uploaded 
+                 (string)
+        title: title of the item for AGOL (string)
+    sde_path: Path to the source .sde connection file
+    temp_dir: Directory for holding reprojected fgdb and .sddraft & .sd files
+    project_path: Path to an existing ArcGIS Pro project
+    map_name: Name of the map in the Pro project to use
+    describe: results of arcpy.da.Describe() on feature class
+
+    returns: path to the .sd file
+    '''
+
+    try:
+        start = datetime.datetime.now()
+
+        sgid_table = os.path.join(sde_path, layer_info['fc_name'])
+        is_table = describe['datasetType'] == 'Table'
+
+        projected_table = project_data(sgid_table, temp_dir, 'tempfgdb.gdb',
+                                       is_table)
+
+        #: Get project and map
+        proj = arcpy.mp.ArcGISProject(project_path)
+        for m in proj.listMaps():
```

Comment:
readability suffers when we use 1 letter variable names.

Reply:
I agree if these variables were used twenty lines away. However, these are 'throwaway' variables that are used only to satisfy arcpy's way of handling things in the next few lines where the context is clear. Furthermore, they avoid stomping over other labels in the namepsace, especially 'map'.

Reply:
selective pep 8 eh? 🤣

Reply:
Not at all: [Foolish consistency is the hobgoblin of little minds.](https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds), just like the lines I let go over 80. If you're done poking fun at my style, I'll get back to real coding.

Reply:
🔥 🚒